File: Codes/Bingran Li/Stack/155. Min Stack.java
class MinStack {
    
    //  how to get the min number in constant time????
    //  just retrieve the min node, do need to go to that node, we only need to know there is a node and its value
    Node head=null; 

    public void push(int x) {
    
       Node myNode = new  Node(x);
       
       if(head == null){
            myNode.min = myNode.val; // didn't consider when head is null, there is no head.min!!!----> PAY ATTENTION 
       } 
       else{
           myNode.min = x<head.min? x:head.min;
       }
       myNode.next = head;
       head = myNode;
       
    }

    public void pop() {
        
        head = head.next;
        
    }

    public int top() {
        
        return head.val;
    }

    public int getMin() {
        
        return head.min;
    }
    
    private class Node{
        
        int val;
        int min;
        Node next;
        
        public Node(int val){
            this.val = val;
            this.next = null;
        }
    }
    
    
}

File: Codes/Yangkun Huang/Stack/155. Min Stack.cpp
stack<int> minstack, min;
void push(int x) {
	if (minstack.empty())
		min.push(x);
	else
		if (x <= min.top())
			min.push(x);
	minstack.push(x);
}

void pop() {
	if (!min.empty())
		if (minstack.top() == min.top())
			min.pop();
	minstack.pop();
}

int top() {
	return minstack.top();
}

int getMin() {
		return min.top();
}
